lib/livebook_web/live/session_live/secrets_component.ex

```diff
@@ -161,26 +268,63 @@ defmodule LivebookWeb.SessionLive.SecretsComponent do
     push_event(socket, "secret_selected", %{select_secret_ref: ref, secret_name: secret_name})
   end
 
-  defp prefill_secret_name(socket) do
-    case socket.assigns.prefill_secret_name do
-      nil ->
-        if unavailable_secret?(socket.assigns.preselect_name, socket.assigns.secrets),
-          do: socket.assigns.preselect_name,
-          else: ""
-
-      prefill ->
-        prefill
-    end
+  defp prefill_secret_name(secret_name, socket) do
+    if unavailable_secret?(
+         secret_name,
+         socket.assigns.secrets,
+         socket.assigns.livebook_secrets
+       ),
+       do: secret_name,
+       else: ""
   end
 
-  defp unavailable_secret?(nil, _), do: false
-  defp unavailable_secret?("", _), do: false
+  defp unavailable_secret?(nil, _, _), do: false
+  defp unavailable_secret?("", _, _), do: false
 
-  defp unavailable_secret?(preselect_name, secrets) do
-    preselect_name not in Enum.map(secrets, & &1.name)
+  defp unavailable_secret?(preselect_name, secrets, livebook_secrets) do
+    preselect_name not in Map.keys(secrets) and preselect_name not in Map.keys(livebook_secrets)
   end
 
   defp title(%{assigns: %{select_secret_ref: nil}}), do: "Add secret"
   defp title(%{assigns: %{select_secret_options: %{"title" => title}}}), do: title
   defp title(_), do: "Select secret"
+
+  defp put_secret(pid, secret, "session"), do: Livebook.Session.put_secret(pid, secret)
+  defp put_secret(_pid, secret, "livebook"), do: Livebook.Secrets.set_secret(secret)
+
+  defp grant_access(secret_name, socket) do
+    secret_value = socket.assigns.livebook_secrets[secret_name]
+    secret = %{name: secret_name, value: secret_value}
+    put_secret(socket.assigns.session.pid, secret, "session")
+  end
+
+  defp livebook_only_secrets(secrets, livebook_secrets) do
+    Enum.reject(livebook_secrets, &(&1 in secrets)) |> Enum.sort()
+  end
+
+  defp maybe_sync_secrets(socket, secret, "livebook") do
```

Comment:
We sync when the user changes the secret within the same section. We should also handle the case where they change the same Livebook secret from another session, no?

Reply:
We only call `maybe_sync_secrets` when the user hits "Save" as far as I understand.

Reply:
Right, this is about sending the saved information to the rest of the system, no?

Reply:
We do `put_secret` here to reflect the new value in the session, unless I miss something :D

Reply:
Sounds good, if we updated the secret when we get the event, it would cover both paths I think.